packetbeat/protos/amqp/amqp.go

```diff
@@ -0,0 +1,883 @@
+			logp.Warn("Failed to get correlation-id in header frame")
+			return true
+		}
+		m.Fields["correlation-id"] = correlationId
+		offset = next
+	}
+
+	if prop1&2 == 2 {
+		replyTo, next, err := getShortString(data, offset+1, uint32(data[offset]))
+		if err {
+			logp.Warn("Failed to get reply-to in header frame")
+			return true
+		}
+		m.Fields["reply-to"] = replyTo
+		offset = next
+	}
+
+	if prop1&1 == 1 {
+		expiration, next, err := getShortString(data, offset+1, uint32(data[offset]))
+		if err {
+			logp.Warn("Failed to get expiration in header frame")
+			return true
+		}
+		m.Fields["expiration"] = expiration
+		offset = next
+	}
+
+	if prop2&128 == 128 {
+		messageId, next, err := getShortString(data, offset+1, uint32(data[offset]))
+		if err {
+			logp.Warn("Failed to get message id in header frame")
+			return true
+		}
+		m.Fields["message-id"] = messageId
+		offset = next
+	}
+
+	if prop2&64 == 64 {
+		t := time.Unix(int64(binary.BigEndian.Uint64(data[offset:offset+8])), 0)
+		m.Fields["timestamp"] = t.Format(AMQPTIMELAYOUT)
+		offset += 8
+	}
+
+	if prop2&32 == 32 {
+		msgType, next, err := getShortString(data, offset+1, uint32(data[offset]))
+		if err {
+			logp.Warn("Failed to get message type in header frame")
+			return true
+		}
+		m.Fields["type"] = msgType
+		offset = next
+	}
+
+	if prop2&16 == 16 {
+		userId, next, err := getShortString(data, offset+1, uint32(data[offset]))
+		if err {
+			logp.Warn("Failed to get user id in header frame")
+			return true
+		}
+		m.Fields["user-id"] = userId
+		offset = next
+	}
+
+	if prop2&8 == 8 {
+		appId, next, err := getShortString(data, offset+1, uint32(data[offset]))
+		if err {
+			logp.Warn("Failed to get app-id in header frame")
+			return true
+		}
+		m.Fields["app-id"] = appId
+		offset = next
+	}
+	return false
+}
+
+func (amqp *Amqp) handleAmqp(m *AmqpMessage, tcptuple *common.TcpTuple, dir uint8) {
+	if amqp.mustHideCloseMethod(m) {
+		return
+	}
+	logp.Debug("amqp", "A message is ready to be handled")
+	m.TcpTuple = *tcptuple
+	m.Direction = dir
+	m.CmdlineTuple = procs.ProcWatcher.FindProcessesTuple(tcptuple.IpPort())
+
+	if m.Method == "basic.publish" {
+		amqp.handlePublishing(m)
+	} else if m.Method == "basic.deliver" || m.Method == "basic.return" ||
+		m.Method == "basic.get-ok" {
+		amqp.handleDelivering(m)
+	} else if m.IsRequest == true {
+		amqp.handleAmqpRequest(m)
+	} else if m.IsRequest == false {
+		amqp.handleAmqpResponse(m)
+	}
+}
+
+func (amqp *Amqp) mustHideCloseMethod(m *AmqpMessage) bool {
+	return amqp.Hide_connection_information == true &&
+		(m.Method == "connection.close" || m.Method == "channel.close") &&
+		getReplyCode(m.Fields) < uint16(300)
+}
+
+func (amqp *Amqp) handleAmqpRequest(msg *AmqpMessage) {
+	// Add it to the HT
+	tuple := msg.TcpTuple
+
+	trans := amqp.getTransaction(tuple.Hashable())
+	if trans != nil {
+		if trans.Amqp != nil {
+			logp.Debug("amqp", "Two requests without a Response. Dropping old request: %s", trans.Amqp)
+		}
```

Comment:
Pipelining is often not defined in protocol spec (e.g. when using TCP), as L4-layer already provides some ordering (messages just buffer up in kernel). It's often a technique used in libs doing async I/O to reduce some latencies.